### PR TITLE
Add example of publishing PyPI package to Local Python repository

### PR DESCRIPTION
**Pull request by jvroberts**
An example of how to make PyPI packages available from an offline Package Manager instance, until Package Manager supports air-gapped/offline PyPI repositories.  

Adds an example bash script that will:
* Download a specified package source tarball from PyPI and its associated binary wheels.
* Upload those package files to a local Package Manager instance using `twine`

### CODE REVIEW

File: pypi-to-local-python/add-pypi-package.sh [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+PACKAGE=$1
+if [ "$2" = "" ]
+then
+    VERSION=.info.version
+else
+    VERSION=\"$2\"
+fi
+
+PKGDIR=$TEMPDIR/$PACKAGE
+mkdir $PKGDIR
```
Comment (jmwoliver) [inline pypi-to-local-python/add-pypi-package.sh:49]: Might want to make this `mkdir -p $PKGDIR` if they use a nested `$TEMPDIR`.

File: pypi-to-local-python/add-pypi-package.sh [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+echo
+echo Downloading package files for "$PACKAGE" from PyPI...
+echo
+
+curl https://pypi.org/pypi/$PACKAGE/json | jq ".releases[$VERSION][] | .url" | xargs --verbose -n1 curl --retry 2 -O --output-dir $PKGDIR
```
Comment (jmwoliver) [inline pypi-to-local-python/add-pypi-package.sh:55]: My local Mac version of `xargs` didn't like the `--verbose` flag. I took it out and it worked though.

File: pypi-to-local-python/add-pypi-package.sh [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+export TWINE_REPOSITORY_URL=$PACKAGEMANAGER_ADDRESS/upload/pypi/$PACKAGEMANAGER_SOURCE
+export TWINE_USERNAME=__token__
+export TWINE_PASSWORD=$PACKAGEMANAGER_TOKEN
+
+echo
+echo Uploading package files to Package Manager...
+echo
+
+python -m twine upload --skip-existing $PKGDIR/*
```
Comment (jmwoliver) [inline pypi-to-local-python/add-pypi-package.sh:69]: My install of `python` uses `python3`, so I ran into an error here. I just changed it in the script, but worth noting that it'll need to be aliased or changed for a customer.

I also have `twine` installed to where I can use it with at the top level (e.g. `twine upload`). Trying to run with `python -m twine upload ...` said it couldn't find it, so also wanted to note that.

**Review by jmwoliver — COMMENTED** [review]
This looks great! I tested it out and made a note of a few things I noticed.

What is the likelihood of an air-gapped customer being able to use this? Would they have `pypi.org` blocked on their network, making them unable to pull down packages like this?

**Comment by shepherdjerred** [comment]
Looks reasonable to me one the issues Jacob brought up are addressed.

File: pypi-to-local-python/add-pypi-package.sh (hunk 2 above)
Comment (jvroberts) [reply pypi-to-local-python/add-pypi-package.sh:55]: Oh, good call.  I had that in for debugging at one point and forgot to take it out.

File: pypi-to-local-python/add-pypi-package.sh (hunk 3 above)
Comment (jvroberts) [reply pypi-to-local-python/add-pypi-package.sh:69]: I can't remember why I switched to `python -m twine` but I agree that calling twine directly is the better way.

**Comment by jvroberts** [comment]
> What is the likelihood of an air-gapped customer being able to use this? Would they have `pypi.org` blocked on their network, making them unable to pull down packages like this?

That's a good question, but technically you could run this from some machine in a DMZ that does have access to PyPI, but also has access to the internal PPM and can upload the packages to there, I guess.  You could also adapt the example to separately pull the packages you want then copy them manually and push them.  It's a starting point at least!

**Review by jonyoder — APPROVED** [review]
This looks really neat! The only suggestion I'd have is to write another alternative script that would let you do the same thing for an entire repo; then a customer could follow this workflow:

1. Install a temporary PPM instance with a trial license (that has access to the internet).
2. Create a Curated-PyPI source/repo and populate it with the packages they need.
3. Run the script against their curated repo to download all the repo packages and upload them to their real PPM server.

**Review by jonyoder — COMMENTED** [review]
Here's a round of some feedback.

File: pypi-to-local-python/add-pypi-package.sh [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+fi
+
+PKGDIR=$TEMPDIR/$PACKAGE
+mkdir -p $PKGDIR
+
+echo
+echo Downloading package files for "$PACKAGE" from PyPI...
+echo
+
+curl https://pypi.org/pypi/$PACKAGE/json | jq ".releases[$VERSION][] | .url" | xargs -n1 curl --retry 2 -O --output-dir $PKGDIR
```
[the remark this answers is not in the document]
Comment (jonyoder) [reply pypi-to-local-python/add-pypi-package.sh:55]: I'd recommend using the `-f` flag to make `curl` fail if the command isn't successful. Then you can check the exit code, like:

```bash
# Get the JSON data from PyPI
url=https://pypi.org/pypi/$PACKAGE/json
json=$(curl -sf $url)
if [[ "$?" -ne 0 ]]; then
  echo "Error downloading package $PACKAGE from $url"
fi

# Download the files
echo $json | jq ".releases[$VERSION][] | .url" | xargs -n1 curl --retry 2 -O --output-dir $PKGDIR
```

[the remark this answers is not in the document]
Comment (jvroberts) [reply pypi-to-local-python/add-pypi-package.sh:55]: Yeah, I originally planned to do that but then the PyPI API was easier.  I'd welcome an adaptation to use P3M instead.

File: pypi-to-local-python/add-pypi-package.sh [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+# Download package files from PyPI #
+####################################
+
+PACKAGE=$1
+if [ "$2" = "" ]
+then
+    VERSION=.info.version
+else
+    VERSION=\"$2\"
+fi
```
[the remark this answers is not in the document]
Comment (jonyoder) [reply pypi-to-local-python/add-pypi-package.sh:46]: My fees are reasonable. :rofl:

[the remark this answers is not in the document]
Comment (jvroberts) [reply pypi-to-local-python/add-pypi-package.sh:46]: Except if $2 is provided, I need it to be double-quoted, but if using the default `.info.version` I need to be unquoted.  What's the cleanest way to make that happen?

[the remark this answers is not in the document]
Comment (jonyoder) [reply pypi-to-local-python/add-pypi-package.sh:46]: I''m sure a real bash wizard would know how, but the smallest I'm coming up with is:
```
VERSION=$([[ -z $2 ]] && echo ".info.version" || echo "\"$2\"")
```

Your original version is more readable!

[the remark this answers is not in the document]
Comment (jonyoder) [reply pypi-to-local-python/add-pypi-package.sh:46]: Actually, here's one idea that's not so bad:

```bash
VERSION=${2:+\"$2\"}
VERSION=${VERSION:-.info.version}
```

[the remark this answers is not in the document]
Comment (jvroberts) [reply pypi-to-local-python/add-pypi-package.sh:46]: That works nicely!

**Pull request merged**